File: integration-test/test-asyncwork.py
"""
测试 异步请求-答复模式
"""

import requests
import random
import json
import time

JAVA_SERVICE = "http://127.0.0.1:7701"
VERTX_SERVICE = "http://127.0.0.1:7703"

create_job_api = JAVA_SERVICE + "/api/asyncwork"
job_detail_api = JAVA_SERVICE + "/api/asyncwork"

HEADERS = {
    "Content-Type": "application/json"
}

def print_response(title, res):
    print(title + ":", res.status_code, res.json())


def create_job():
    n = random.randint(1, 1000)
    request_data = {
        "name": "name-" + str(n),
        "value": "value-" + str(n)
    }
    res = requests.post(create_job_api, json.dumps(request_data), headers=HEADERS)
    data = res.json()
    if res.status_code == 202:
        print("Create job:", data["jobId"])
        return data
    else:
        print("Create job failed:", res.status_code, data)
        raise Exception("Create job failed")

def check_job_state(state_endpoint):
    url = JAVA_SERVICE + state_endpoint
    i = 0
    while i < 20:
        i += 1
        res = requests.get(url)
        if res.ok:
            data = res.json()
            print("Job state:", data["state"])
            if data["state"] in ["RUNNING", "NEW"]:
                time.sleep(1.0)
            else:
                return data
        else:
            print("Job state failed:", res.status_code, res.json())
            raise Exception("Check job state failed")
    raise Exception("Retry timeout")

def check_job_result(result_endpoint):
    url = JAVA_SERVICE + result_endpoint
    res = requests.get(url)
    data = res.json()

    if res.ok:
        print("Job result:", data["result"])
    else:
        print("Job result failed:", res.status_code, data)
        raise Exception("Check job result failed")

if __name__ == "__main__":
    
    data = create_job()
    data = check_job_state(data["stateEndpoint"])

    if "resultEndpoint" in data:
        check_job_result(data["resultEndpoint"])

File: integration-test/test-asyncwork.js
/**
 * 测试 异步请求-答复模式
 */

import { check, sleep } from "k6"
import http from "k6/http"

const java_service = "http://localhost:7701"
const vertx_service = "http://localhost:7703"

export const options = {
};

const headers = {
  'Content-Type': 'application/json' 
}

const printResponse = (title, res) => {
  console.log(title + " :", res.status, res.json())
}


export default function () {

  let n = Math.floor(Math.random() * 1000)
  const data = {
    "name": "TestJob-" + n,
    "value": "value-" + n
  }
  const res = http.post(`${java_service}/api/asyncwork`, JSON.stringify(data), {headers: headers})

  printResponse("Create Job", res)

  if (res.status >= 300 ) {
    return
  }

  let stateEndpoint = res.json().stateEndpoint
  let retryAfter = res.json().retryAfter

  let resultEndpoint

  while(true) {
    const res2 = http.get(java_service + stateEndpoint)
    printResponse("Job State", res2)
    const data = res2.json()
    if (data.state == "NEW" || data.state == "RUNNING") {
      sleep(1)
    } else if (data.state == "FINISH") {
      resultEndpoint = data.resultEndpoint
      break
    } else {
      break
    }
  }

  if (resultEndpoint) {
    const res3 = http.get(java_service + resultEndpoint)
    printResponse("Job Result", res3)
  }

}
